include/margo-monitoring.h

```diff
@@ -0,0 +1,397 @@
+    /* input */
+    hg_bulk_t handle;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_rpc_handler_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    /* output */
+    ABT_pool    pool;
+    hg_return_t ret;
+};
+
+struct margo_monitor_rpc_ult_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    /* output */
+};
+
+struct margo_monitor_wait_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    margo_request request;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_sleep_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    double timeout_ms;
+    /* output */
+};
+
+struct margo_monitor_set_input_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t   handle;
+    margo_request request;
+    const void*   data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_set_output_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t   handle;
+    margo_request request;
+    const void*   data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_get_input_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_get_output_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_free_input_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_free_output_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    hg_handle_t handle;
+    const void* data;
+    /* output */
+    hg_return_t ret;
+};
+
+struct margo_monitor_prefinalize_args {
+    margo_monitor_data_t uctx;
+};
+
+struct margo_monitor_finalize_args {
+    margo_monitor_data_t uctx;
+};
+
+struct margo_monitor_cb_args {
+    margo_monitor_data_t uctx;
+    /* input */
+    const struct hg_cb_info* info;
+    margo_request            request;
+    /* output */
+    hg_return_t ret;
+};
+
+/**
+ * @brief Set a monitor structure for Margo to use. The structure will be
+ * internally copied and the user may free the input argument after the call.
+ * Passing NULL as monitor is valid and will disable monitor altogether.
+ *
+ * Note: if a monitor is already in place, its finalize function pointer will
+ * be called before the new monitor is installed. If provided, The initialize
+ * function of the new monitor will be called and the monitor's uargs will be
+ * replaced internally with the returned value of the initialize call.
+ *
+ * @param mid Margo instance
+ * @param monitor Monitor structure
+ * @param config JSON-formatted configuration or NULL
+ *
+ * @return HG_SUCCESS or other error code
+ */
+hg_return_t margo_set_monitor(margo_instance_id           mid,
```

Comment:
I haven't thought this through, but if how hard would it be to stack monitors?  We shouldn't pursue it if it were disruptive, but if there were something were the onus were on the monitor implementor maybe.  Imagine I wanted a new monitor that did something peculiar to look at bulk create statistics, but I also wanted to keep a "normal" monitor too. Maybe I could use a get_monitor call or similar to get the existing pointers so that my bulk_create handlers could just relay to the existing monitor's fn pointer after capturing what it wanted?  If there is something simple to accomplish that we could do it now, if not then let's skip that feature until we need it.  My use case is conjecture right now.

Reply:
I thought about it (for a few hours, actually). The problem with stacking monitors is keeping states across START and END calls. As you may have seen, all the structures have a field that can be used for the monitor to place data. With stacked monitors, it becomes much more difficult to manage.

However this is something that can be done through aggregation, rather than stacking, and Bedrock would be the right place for that. Bedrock would inject its own monitoring implementation into margo, and this monitoring implementation could allow multiple "backend" implementations to be invoked (and also dlopen-ed from shared libraries, etc.), swapping the content of the user field as needed in between invokation of each monitor.

This is also why I didn't implement a way of loading a monitor from a dynamic library directly in Margo. This is something that is better suited to Bedrock.

Reply:
Ok, I see.  I tvote to go with what you have for now then (no stacking).  We can revisit later or do something at the bedrock level when we have a use case.